samples/boards/st/ble_802154/ble_hr_802154_echo_client/src/udp.c

```diff
@@ -0,0 +1,382 @@
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(ble_hr_802154_echo_client_sample, LOG_LEVEL_DBG);
+
+#include <zephyr/kernel.h>
+#include <errno.h>
+#include <stdio.h>
+
+#include <zephyr/posix/sys/socket.h>
+#include <zephyr/posix/arpa/inet.h>
+#include <zephyr/posix/unistd.h>
+
+#include <zephyr/net/socket.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/random/random.h>
+
+#include "common.h"
+
+#if defined(CONFIG_NET_UDP)
+
+#define RECV_BUF_SIZE 1280
+#define UDP_SLEEP K_MSEC(150)
+#define UDP_WAIT K_SECONDS(10)
+
+static APP_BMEM char recv_buf[RECV_BUF_SIZE];
+
+static K_THREAD_STACK_DEFINE(udp_tx_thread_stack, UDP_STACK_SIZE);
+static struct k_thread udp_tx_thread;
+
+/* Kernel objects should not be placed in a memory area accessible from user
+ * threads.
+ */
+static struct udp_control udp4_ctrl, udp6_ctrl;
+static struct k_poll_signal udp_kill;
+
+static int send_udp_data(struct sample_data *data);
+static void wait_reply(struct k_timer *timer);
+static void wait_transmit(struct k_timer *timer);
+
+static void process_udp_tx(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	struct k_poll_event events[] = {
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &udp_kill),
+#if defined(CONFIG_NET_IPV4)
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &udp4_ctrl.tx_signal),
+#endif
+#if defined(CONFIG_NET_IPV6)
+		K_POLL_EVENT_INITIALIZER(K_POLL_TYPE_SIGNAL,
+					 K_POLL_MODE_NOTIFY_ONLY,
+					 &udp6_ctrl.tx_signal),
+#endif
+	};
+
+	while (true) {
+		k_poll(events, ARRAY_SIZE(events), K_FOREVER);
+
+		for (int i = 0; i < ARRAY_SIZE(events); i++) {
+			unsigned int signaled;
+			int result;
+
+			k_poll_signal_check(events[i].signal, &signaled, &result);
+			if (signaled == 0) {
+				continue;
+			}
+
+			k_poll_signal_reset(events[i].signal);
+			events[i].state = K_POLL_STATE_NOT_READY;
+
+			if (events[i].signal == &udp_kill) {
+				return;
+			} else if (events[i].signal == &udp4_ctrl.tx_signal) {
+				send_udp_data(&conf.ipv4);
+			} else {
+				if (events[i].signal == &udp6_ctrl.tx_signal) {
+					send_udp_data(&conf.ipv6);
+				}
+			}
+		}
+	}
+}
+
+static void udp_control_init(struct udp_control *ctrl)
+{
+	k_timer_init(&ctrl->rx_timer, wait_reply, NULL);
+	k_timer_init(&ctrl->tx_timer, wait_transmit, NULL);
+	k_poll_signal_init(&ctrl->tx_signal);
+}
+
+static void udp_control_access_grant(struct udp_control *ctrl)
+{
+	k_thread_access_grant(k_current_get(),
+			      &ctrl->rx_timer,
+			      &ctrl->tx_timer,
+			      &ctrl->tx_signal);
+}
+
+void init_udp(void)
+{
+	/* k_timer_init() is not a system call, therefore initialize kernel
+	 * objects here.
+	 */
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		udp_control_init(&udp4_ctrl);
+		conf.ipv4.udp.ctrl = &udp4_ctrl;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		udp_control_init(&udp6_ctrl);
+		conf.ipv6.udp.ctrl = &udp6_ctrl;
+	}
```

Comment:
Do we intent to support both configurations ?
I can see the following in prj.conf:
```
CONFIG_NET_IPV6=y
CONFIG_NET_IPV4=n
```